File: py/aoc2024/day12.py
"""
Day 12: Garden Groups
"""

from collections import Counter, defaultdict
import textwrap
from typing import Generator

SAMPLE_INPUT_1 = textwrap.dedent(
    """\
    AAAA
    BBCD
    BBCC
    EEEC
    """
)
SAMPLE_INPUT_2 = textwrap.dedent(
    """\
    OOOOO
    OXOXO
    OOOOO
    OXOXO
    OOOOO
    """
)
SAMPLE_INPUT_3 = textwrap.dedent(
    """\
    RRRRIICCFF
    RRRRIICCCF
    VVRRRCCFFF
    VVRCCCJFFF
    VVVVCJJCFE
    VVIVCCJJEE
    VVIIICJJEE
    MIIIIIJJEE
    MIIISIJEEE
    MMMISSJEEE
    """
)
SAMPLE_INPUT_4 = textwrap.dedent(
    """\
    EEEEE
    EXXXX
    EEEEE
    EXXXX
    EEEEE
    """
)
SAMPLE_INPUT_5 = textwrap.dedent(
    """\
    AAAAAA
    AAABBA
    AAABBA
    ABBAAA
    ABBAAA
    AAAAAA
    """
)


def _groups(data: str) -> Generator[set[tuple[int, int]]]:
    lines = data.splitlines()
    visited = set()
    for y, line in enumerate(lines):
        for x, char in enumerate(line):
            group = set()
            stack = [(y, x)]
            while stack:
                y2, x2 = pos = stack.pop()
                if lines[y2][x2] != char or pos in visited:
                    continue
                visited.add(pos)
                group.add(pos)
                if 0 < y2:
                    stack.append((y2 - 1, x2))
                if 0 < x2:
                    stack.append((y2, x2 - 1))
                if x2 < len(lines[y2]) - 1:
                    stack.append((y2, x2 + 1))
                if y2 < len(lines) - 1:
                    stack.append((y2 + 1, x2))
            if group:
                yield (group)


def _perimeter1(group: set[tuple[int, int]]) -> int:
    edges = Counter()
    for y, x in group:
        edges.update(
            (
                (2 * y - 1, 2 * x),
                (2 * y, 2 * x - 1),
                (2 * y, 2 * x + 1),
                (2 * y + 1, 2 * x),
            )
        )
    return sum(value == 1 for value in edges.values())


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT_1)
    140
    >>> part1(SAMPLE_INPUT_2)
    772
    >>> part1(SAMPLE_INPUT_3)
    1930
    """
    return sum(len(group) * _perimeter1(group) for group in _groups(data))


def _perimeter2(group: set[tuple[int, int]]) -> int:
    edges = defaultdict(set)
    for y, x in group:
        edges[(2 * y - 1, 2 * x)].add(0)
        edges[(2 * y, 2 * x - 1)].add(1)
        edges[(2 * y, 2 * x + 1)].add(3)
        edges[(2 * y + 1, 2 * x)].add(2)
    lines = defaultdict(set)
    for (y, x), value in edges.items():
        if len(value) == 1:
            value = next(iter(value))
            lines[x if value % 2 else y + 1].add((x + y, value))
    return sum(
        1 + sum(abs(q - p) > 2 or b != d for (q, b), (p, d) in zip(line, line[1:]))
        for line in map(sorted, lines.values())
    )


def part2(data: str) -> int:
    """
    >>> part2(SAMPLE_INPUT_1)
    80
    >>> part2(SAMPLE_INPUT_2)
    436
    >>> part2(SAMPLE_INPUT_4)
    236
    >>> part2(SAMPLE_INPUT_5)
    368
    >>> part2(SAMPLE_INPUT_3)
    1206
    """
    return sum(len(group) * _perimeter2(group) for group in _groups(data))


parts = (part1, part2)

File: rs/src/day12.rs
use std::collections::{BTreeMap, BTreeSet};

use if_chain::if_chain;
use itertools::Itertools;

struct Data<'a> {
    lines: Vec<&'a str>,
}

impl Data<'_> {
    fn new(data: &str) -> Data {
        Data {
            lines: data.lines().collect(),
        }
    }

    fn groups(&self) -> impl Iterator<Item = BTreeSet<(usize, usize)>> + use<'_> {
        self.lines
            .iter()
            .enumerate()
            .flat_map(|(y, line)| line.bytes().enumerate().map(move |(x, b)| ((y, x), b)))
            .scan(BTreeSet::new(), |visited, (pos, b)| {
                if !visited.insert(pos) {
                    return Some(None);
                }
                let mut group = BTreeSet::new();
                let mut stack = vec![pos];
                while let Some(pos @ (y, x)) = stack.pop() {
                    group.insert(pos);
                    for (dy, dx) in [(-1, 0), (0, -1), (0, 1), (1, 0)] {
                        let pos @ (y, x) = (y.wrapping_add_signed(dy), x.wrapping_add_signed(dx));
                        if_chain! {
                            if let Some(line) = self.lines.get(y);
                            if line.as_bytes().get(x) == Some(&b);
                            if visited.insert(pos);
                            then {
                                stack.push(pos);
                            }
                        }
                    }
                }
                Some(Some(group))
            })
            .flatten()
    }
}

fn perimeter1(group: &BTreeSet<(usize, usize)>) -> usize {
    let mut edges = BTreeMap::<_, u8>::new();
    for (y, x) in group.iter() {
        for edge in [
            (2 * y, 2 * x + 1),
            (2 * y + 1, 2 * x),
            (2 * y + 1, 2 * x + 2),
            (2 * y + 2, 2 * x + 1),
        ] {
            *edges.entry(edge).or_default() += 1;
        }
    }
    edges.into_values().filter(|value| *value == 1).count()
}

pub fn part1(data: &str) -> usize {
    Data::new(data)
        .groups()
        .map(|group| group.len() * perimeter1(&group))
        .sum()
}

fn perimeter2(group: &BTreeSet<(usize, usize)>) -> usize {
    let mut edges = BTreeMap::<_, u8>::new();
    for (y, x) in group.iter() {
        *edges.entry((2 * y, 2 * x + 1)).or_default() |= 1;
        *edges.entry((2 * y + 1, 2 * x)).or_default() |= 2;
        *edges.entry((2 * y + 1, 2 * x + 2)).or_default() |= 8;
        *edges.entry((2 * y + 2, 2 * x + 1)).or_default() |= 4;
    }
    let mut lines = BTreeMap::<_, BTreeSet<_>>::new();
    for ((y, x), n) in edges {
        if n & (n - 1) == 0 {
            lines
                .entry(if n.ilog2() % 2 == 0 { y + 1 } else { x })
                .or_default()
                .insert((x + y, n));
        }
    }
    lines
        .into_values()
        .map(|line| {
            line.into_iter()
                .tuple_windows()
                .filter(|((p, b), (q, d))| p.abs_diff(*q) > 2 || b != d)
                .count()
                + 1
        })
        .sum()
}

pub fn part2(data: &str) -> usize {
    Data::new(data)
        .groups()
        .map(|group| group.len() * perimeter2(&group))
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE_1: &str = indoc! {"
        AAAA
        BBCD
        BBCC
        EEEC
    "};

    static EXAMPLE_2: &str = indoc! {"
        OOOOO
        OXOXO
        OOOOO
        OXOXO
        OOOOO
    "};

    static EXAMPLE_3: &str = indoc! {"
        RRRRIICCFF
        RRRRIICCCF
        VVRRRCCFFF
        VVRCCCJFFF
        VVVVCJJCFE
        VVIVCCJJEE
        VVIIICJJEE
        MIIIIIJJEE
        MIIISIJEEE
        MMMISSJEEE
    "};

    static EXAMPLE_4: &str = indoc! {"
        EEEEE
        EXXXX
        EEEEE
        EXXXX
        EEEEE
    "};

    static EXAMPLE_5: &str = indoc! {"
        AAAAAA
        AAABBA
        AAABBA
        ABBAAA
        ABBAAA
        AAAAAA
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(140, part1(EXAMPLE_1));
        assert_eq!(772, part1(EXAMPLE_2));
        assert_eq!(1930, part1(EXAMPLE_3));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(80, part2(EXAMPLE_1));
        assert_eq!(436, part2(EXAMPLE_2));
        assert_eq!(236, part2(EXAMPLE_4));
        assert_eq!(368, part2(EXAMPLE_5));
        assert_eq!(1206, part2(EXAMPLE_3));
    }
}
